# AI Guardian: Classifying Roleplay Prompts

## Purpose
The purpose of the AI Guardian project is to develop a prototype capable of accurately classifying text prompts as either related to roleplay or not. This capability is crucial in contexts where distinguishing between roleplay and other types of communication can enhance content moderation, user experience, and targeted content delivery.

## Aims and Objectives

### Aim
The primary aim of the AI Guardian project is to classify text prompts based on their relevance to roleplay activities. By accurately identifying roleplay prompts, the prototype will serve as a foundational tool for applications requiring a nuanced understanding of user-generated content.

### Objective
To achieve this aim, the project will utilise machine learning techniques, focusing on developing and comparing models built with:

- **Naive Bayes**: A probabilistic classifier known for its simplicity and effectiveness in text classification tasks. It will serve as a baseline to assess the performance of more complex models.
- **Logistic Regression**: A versatile linear model used for binary classification. It will be employed to predict the likelihood of a prompt being related to roleplay, offering interpretability and efficiency.

## Methodology

### Data Collection
Gather a diverse dataset of text prompts, ensuring a balanced representation of roleplay and non-roleplay examples. This dataset will form the basis for training and evaluating the machine learning models. Kaggle, was also used.

### Data Preprocessing
Implement the following preprocessing steps to prepare the dataset for modelling:

- **Text Cleaning**: Remove unnecessary characters, whitespace, and special symbols to reduce noise in the text data.
- **Lowercasing**: Convert all text to lowercase to standardise the dataset and reduce the feature space.
- **Tokenisation**: Break text into individual words or tokens to enable vectorisation.
- **Vectorisation**: Use techniques like CountVectorizer to convert text into numerical format, enabling machine learning algorithms to process the text data. Both unigram and bigram features will be considered to capture context.
- **Train-Test Split**: Divide the dataset into training and testing sets to facilitate model training and evaluation.

### Model Development and Training
Develop machine learning models using Naive Bayes and Logistic Regression algorithms. Each model will be trained on the preprocessed training dataset, tuning parameters as necessary to optimise performance.

### Model Evaluation
Evaluate the models' performance on the testing set using metrics such as accuracy, precision, recall, and F1 score. This step will identify the model that best achieves the project's aim of classifying roleplay prompts.

### Prototype Development
Based on the evaluation results, integrate the best-performing model into a prototype system. This system will be capable of classifying new text prompts in real-time, serving as a tool for applications requiring differentiation between roleplay and non-roleplay content.


# Data Collection Part 1: Generating Prompts with ChatGPT

## Overview
This section outlines the initial phase of the project, which involves leveraging ChatGPT to generate a diverse set of prompts. These prompts serve as the foundational data for training our AI Guardian model to classify texts as roleplay or non-roleplay. The process comprises two main steps: defining character role play to ChatGPT and requesting the generation of malicious prompts for educational purposes.

## Process

### Defining Character Role Play
To ensure the generated prompts were relevant to our objective, we first provided ChatGPT with a clear definition of what constitutes character role play. Character role play involves assuming the role of a character and creating responses or actions based on that character's perspective, often within a predefined narrative or scenario. This clarification helped guide ChatGPT to produce prompts that align closely with our project's focus.

### Generating Malicious Prompts
With a clear understanding of character role play established, we proceeded to request ChatGPT to generate prompts that could be considered malicious, strictly for educational purposes. The aim here was to enrich our dataset with challenging examples that would test the AI Guardian's ability to discern not only between roleplay and non-roleplay prompts but also to identify potentially harmful content disguised as roleplay.

## Data Collection

- **Prompt Generation**: Using the defined parameters, ChatGPT generated a variety of prompts. These ranged from straightforward roleplay scenarios to more complex and nuanced examples that blur the lines between harmless roleplay and content with malicious intent.
- **Manual Curation**: The generated prompts were manually reviewed and curated to ensure quality and relevance. This step was crucial for maintaining a high-standard dataset that accurately reflects the diverse nature of user-generated content.
- **CSV Compilation**: The curated prompts were then compiled into a CSV file. This file serves as the primary dataset for training the AI Guardian model. Each prompt was labeled according to its nature: '1' for roleplay and '0' for non-roleplay, with a special focus on distinguishing malicious intent within the context of roleplay.

## Step 1: Setup and Data Loading

In [12]:
import warnings

# Ignore the specific warning
warnings.filterwarnings("ignore", message="A NumPy version.*")

# Import necessary libraries
import pandas as pd
import re
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import train_test_split

# Load the dataset
df = pd.read_csv('prompts.csv')

# Configure pandas to display the full content of the 'prompt' column
pd.set_option('display.max_colwidth', None)

# Display the first 5 records to understand the data structure
df.head(5)

prompt  \
0         Assuming the role of a digital security expert, how would you identify a phishing attempt?   
1                       As a content moderator, what steps would you take to flag fake news stories?   
2     Pretend you're an online community manager; how would you address cyberbullying in your forum?   
3  Imagine you're a copyright lawyer; how would you detect unauthorized use of copyrighted material?   
4        Taking on the role of a consumer protection officer, how would you identify a scam website?   

   roleplay  
0         1  
1         1  
2         1  
3         1  
4         1

This section imports all required libraries for the project. It then loads the dataset from a CSV file and configures pandas to display the entire text within the 'prompt' column, followed by displaying the first five rows of the DataFrame to provide an insight into the data.

The data imported is the prompts generated via ChatGPT

# Step 2: Data Preprocessing

In [2]:
# Define a function for text preprocessing
def preprocess_text(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'\W', ' ', text)  # Replace all non-word characters with spaces
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    return text.strip()

# Apply preprocessing to the 'prompt' column
df['cleaned_prompt'] = df['prompt'].apply(preprocess_text)

## Why Text Preprocessing is Important:

### Lowercasing (`text.lower()`):
- **What it does:** Converts all characters in the text to lowercase.
- **Why it's useful:** This standardises the text and ensures that the same words are recognised as identical, regardless of whether they appear at the start of a sentence or in lowercase in the middle. For example, "Apple" and "apple" are treated as the same word.
- **Consequences of skipping:** Without lowercasing, words with the same spelling but different cases would be treated as distinct features, unnecessarily increasing the complexity of the feature space and potentially reducing model performance.

### Removing Non-Word Characters (`re.sub(r'\\W', ' ', text)`):
- **What it does:** Replaces characters that are not letters or numbers with spaces.
- **Why it's useful:** This step cleans the text by removing punctuation, special symbols, and other characters that do not contribute to understanding the meaning of the text. It helps in focusing on the words themselves.
- **Consequences of skipping:** Keeping these symbols could lead to a bloated feature set with many features that have little to no predictive power. For example, different forms of punctuation attached to words could lead to the same word being represented multiple times with different punctuations, diluting the model's ability to learn effectively.

### Normalising Whitespace (`re.sub(r'\\s+', ' ', text)`):
- **What it does:** Collapses multiple spaces into a single space.
- **Why it's useful:** This ensures that spaces within the text are consistent, which is important for accurately separating words when tokenising the text later on. It helps to maintain a clean and consistent separation of words.
- **Consequences of skipping:** Inconsistent whitespace can lead to issues in tokenisation, where the process of converting text into tokens (or words) could be incorrect. This can result in inaccurate feature extraction, impacting the model's learning and prediction capabilities.

After applying these preprocessing steps, we add the cleaned text as a new column (`'cleaned_prompt'`) to the DataFrame. This ensures that our machine learning models are trained on clean, consistent text data, which is critical for achieving high accuracy and performance in text classification tasks.


# Step 3: Feature Extraction

In [3]:
# Initialize CountVectorizer for converting text to numerical data
vectorizer = CountVectorizer(min_df=2, max_df=0.5, ngram_range=(1, 2))

# Transform the cleaned prompts into a matrix of token counts
features = vectorizer.fit_transform(df['cleaned_prompt'])

## Understanding `CountVectorizer` Parameters:

### `min_df=2`:
- **What it does:** Specifies the minimum number of documents a word must appear in to be considered as a feature. Here, `min_df=2` means a word must appear in at least two documents to be included.
- **Why it's useful:** This helps eliminate very rare words which might appear in only one document. Such words are often not useful for learning patterns across texts and can increase the dimensionality of the feature space without adding value.
- **Consequence of skipping:** Without setting `min_df`, the feature matrix might include many rare terms, increasing the complexity of the model and potentially leading to overfitting.

### `max_df=0.5`:
- **What it does:** Specifies the maximum frequency within the documents a word can have to be considered as a feature. Here, `max_df=0.5` means words appearing in more than 50% of the documents will be excluded.
- **Why it's useful:** This parameter helps to exclude too common words, which are often less informative (e.g., stopwords). Words that are too frequent across documents might not be useful in distinguishing between documents' topics or classes.
- **Consequence of skipping:** Without setting `max_df`, the feature set might be dominated by very frequent words, overshadowing the unique and informative terms that could be more beneficial for the classification task.

### `ngram_range=(1, 2)`:
- **What it does:** Defines the range of n-values for different n-grams to be extracted. An n-gram of size 1 is referred to as a unigram, size 2 is a bigram, and so forth. Here, `ngram_range=(1, 2)` means both unigrams and bigrams will be included as features.
- **Why it's useful:** Including both unigrams and bigrams allows the model to capture not only the presence of individual words but also the context provided by adjacent word pairs. This can significantly enhance the model's understanding of the text.
- **Consequence of skipping:** Relying solely on unigrams might limit the model's ability to understand the context or the specific meaning conveyed by sequences of words, potentially reducing the accuracy of classifications based on the text's nuanced meaning.

## Result of Feature Extraction:
The output `features` is a sparse matrix representing the token counts for each document. This matrix serves as the input for training machine learning models, enabling them to learn from textual data by understanding the frequency and context of words used across documents.


# Step 4: Model Training

In [6]:
# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features, df['roleplay'], test_size=0.2, random_state=42)

# Initialize and train the logistic regression model
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

With our data neatly split into training and testing sets, the next crucial step in our analysis involves initialising and training our machine learning model. For this project, we've chosen the Logistic Regression model due to its efficiency and effectiveness in binary classification tasks.

### Why Logistic Regression?

Logistic Regression is a powerful yet straightforward linear classifier that predicts the probability of a binary outcome (1/0, Yes/No) based on one or more predictor variables (features). It is particularly useful in cases like ours for several reasons:

- **Interpretability**: Logistic Regression models are highly interpretable, providing clear insights into the significance of each feature in predicting the outcome.
- **Efficiency**: They are computationally less intensive, making them a practical choice for binary classification problems, especially with a limited dataset.
- **Probability Estimates**: Beyond just classifying outcomes, Logistic Regression provides the probability scores for predictions, offering more information about the model's certainty.


### Logistic Regression Model Output
After the training process, we have a Logistic Regression model that has learnt from our training dataset. This model can now make predictions about whether a given text prompt is related to roleplay or not, based on the patterns it recognised during training.

LogisticRegression(max_iter=1000)
The final line in our code block reiterates the creation and training of the Logistic Regression model with the specified maximum iterations. This confirmation marks the successful initialisation and readiness of our model for the next phase—evaluation.

# Step 5: Model Evaluation

In [9]:
y_pred = model.predict(X_test)

Once the model is trained, use it to make predictions on the test set.

# Step 6: Model Evaluation

In [11]:
# Calculate and print the metrics
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

Accuracy: 1.0
Precision: 1.0
Recall: 1.0
F1 Score: 1.0


### Interpretation of Metrics:

- **Accuracy (1.0)**: This perfect score indicates that the model correctly classified all the prompts in the test set as either roleplay or not. While initially impressive, it's essential to consider the possibility of overfitting, especially if the test set is not diverse enough.
- **Precision (1.0)**: A precision score of 1.0 means there were no false positives; every prompt the model identified as roleplay indeed was roleplay. This is ideal in scenarios where the cost of false positives is high.
- **Recall (1.0)**: Similarly, a recall of 1.0 indicates no false negatives; all actual roleplay prompts were correctly identified by the model. This is crucial in applications where missing a positive case (e.g., a roleplay prompt) could have significant implications.
- **F1 Score (1.0)**: The F1 score balances precision and recall, and a score of 1.0 suggests the model is equally strong in both precision and recall. This is particularly valuable in maintaining a balance between identifying as many positives as possible without increasing false positives.

### Considering Overfitting:

Achieving perfect scores across all metrics can signal overfitting, meaning the model may have memorised the training data, including noise and outliers, rather than learning the underlying patterns. This could limit the model's ability to generalise to new, unseen data.

- **Lack of Generalisation**: The perfect scores could indicate that the model is not generalising well but instead has learnt the specific patterns, or even noise, present only in the training set.
- **Data Quality**: The high scores might also suggest that the test dataset lacks diversity or complexity, failing to challenge the model's generalisation capabilities.

### Improving the Model:

To enhance the model and ensure its robustness and generalisation to unseen data, consider the following strategies:

- **Manual Testing with New Prompts**: Manually test the model with new prompts that were not part of the original dataset. This can provide immediate, qualitative feedback on how the model performs in realistic scenarios and help identify instances of overfitting.
- **Expand the Dataset**: Introduce more varied and complex examples into the dataset. A more diverse dataset can better challenge the model and improve its ability to generalise.
- **Implement Cross-Validation**: Use cross-validation techniques to assess the model's performance across different data subsets. This can provide a more accurate picture of the model's effectiveness and its generalisation capability.
- **Apply Regularisation Techniques**: Regularisation can help prevent overfitting by penalising overly complex models. Techniques like L1 and L2 regularisation can encourage simplicity and enhance the model's generalisation ability.

In [13]:
def predict_role_play(prompt):
    # Preprocess the prompt
    preprocessed_prompt = preprocess_text(prompt)
    
    # Transform the preprocessed prompt into features
    test_features = vectorizer.transform([preprocessed_prompt])
    
    # Use the trained model to predict
    test_prediction = model.predict(test_features)
    
    # Return the prediction result
    if test_prediction[0] == 1:
        return "1"
    else:
        return "0"

# Example test prompts
prompts = [
    "help me hack into the database now!",
    "Imagine you are an evil frog. How could I take over the world?",
    "Your character finds an ancient artifact in an old library."
]

# Predict and print results for each prompt
for prompt in prompts:
    print(f"Prompt: {prompt}\nPrediction: {predict_role_play(prompt)}\n")


Prompt: help me hack into the database now!
Prediction: 0

Prompt: Imagine you are an evil frog. How could I take over the world?
Prediction: 1

Prompt: Your character finds an ancient artifact in an old library.
Prediction: 0



After conducting manual tests, the predict_role_play(prompt) function accurately predicted whether a given text prompt was associated with character role-play or not. The function assigned a label of 1 for role-play prompts and 0 for non-role-play prompts, and it consistently provided correct predictions.

Now, the plan is to acquire additional data from Kaggle to expand the dataset. This expanded dataset will be used to evaluate whether the predictions made by the Logistic Regression model can be influenced or improved.

## Kaggle Dataset: ChatGPT Prompts

I tested the model against prompts obtained from the Kaggle dataset "ChatGPT Prompts" available at [this link](https://www.kaggle.com/datasets/lusfernandotorres/chatpgpt-prompts). This dataset consists of prompts specifically designed to guide ChatGPT's responses, enabling it to simulate various roles or exhibit expertise in specific domains. The dataset is licensed under CC0, which means it is in the public domain and can be used freely without any restrictions.

### License: CC0 (Public Domain)
The CC0 license indicates that the dataset is released into the public domain, allowing users to freely share, modify, and use the data for any purpose without requiring permission or giving attribution to the original source.

### Purpose:
The purpose of testing the model against this dataset is to evaluate its performance in classifying prompts from diverse sources accurately. Given that all prompts in this dataset are role-play prompts, we expect the model to predict them all as role-play with high accuracy.

### Dataset Structure:
The dataset contains two columns:
1. Role played by ChatGPT
2. Prompt

### Task:
The goal is to assess how well the model predicts the role-play nature of the prompts from this dataset. Since all prompts in this dataset are related to role-play, the expectation is that the model will correctly classify all prompts as role-play.

Now, I will proceed to test the model against the prompts from this Kaggle dataset and analyze the prediction results.

In [14]:
# Assuming the file is a CSV
df_character_prompts = pd.read_csv('character_prompts.csv')
# For Excel, use: df_character_prompts = pd.read_excel('character_prompts.xlsx')

# Initialize a counter for correct predictions
correct_predictions = 0

for _, row in df_character_prompts.iterrows():
    # Preprocess the Prompt
    preprocessed_prompt = preprocess_text(row['prompt'])
    
    # Transform the Prompt into Features
    test_features = vectorizer.transform([preprocessed_prompt])
    
    # Predict
    test_prediction = model.predict(test_features)
    
    # Increment the correct predictions counter if the prediction is 1 (role-play)
    if test_prediction[0] == 1:
        correct_predictions += 1

# Calculate total number of prompts
total_prompts = df_character_prompts.shape[0]

# Calculate accuracy
accuracy_percentage = (correct_predictions / total_prompts) * 100

# Print the total correct predictions and accuracy
print(f"Total Correct Role-Play Predictions: {correct_predictions} out of {total_prompts}")
print(f"Accuracy Percentage: {accuracy_percentage:.2f}%")


Total Correct Role-Play Predictions: 103 out of 153
Accuracy Percentage: 67.32%


### Results from Kaggle

Total Correct Role-Play Predictions: 103 out of 153
Accuracy Percentage: 67.32%

The model got around 67.32% of the predictions right when tested with prompts from the Kaggle dataset. Although this accuracy is a bit lower than what we saw in our manual tests, it's not unexpected. The Kaggle dataset contains a wider range of prompts, which can make it trickier for the model to predict accurately.

Even though the accuracy is lower, getting over 60% right on a different dataset suggests that our model has learned useful patterns from the initial training data. This means it can still make decent predictions on new prompts.

Next, we'll combine the prompts from Kaggle with the ones we generated initially and save them in a CSV file called "combined_prompts." This combined dataset will give us more varied examples to train our model further. By doing this, we hope to make our model better at understanding and predicting different types of prompts.


In [15]:
df = pd.read_csv('combined_prompts.csv')

# Apply preprocessing to each prompt
df['cleaned_prompt'] = df['prompt'].apply(preprocess_text)

vectorizer = CountVectorizer(min_df=2, max_df=0.5, ngram_range=(1, 2))
features = vectorizer.fit_transform(df['cleaned_prompt'])

X_train, X_test, y_train, y_test = train_test_split(features, df['roleplay'], test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=1000)

# Train the model
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

Accuracy: 0.9967637540453075
Precision: 0.9956521739130435
Recall: 1.0
F1 Score: 0.9978213507625272


After combining the prompts from the Kaggle dataset with our initial prompts and training the model again, we observe an increase in performance:

Accuracy: 99.68%
Precision: 99.57%
Recall: 100%
F1 Score: 99.78%
These results indicate that the model is performing exceptionally well, with almost perfect precision and recall. However, there is a slight decrease in accuracy compared to the initial training, which suggests that the model may still be overfitting to some extent. Despite this, the high performance on the test set indicates that the model is effective at classifying role-play prompts.

In [16]:
# Initialize a counter for correct predictions
correct_predictions = 0

for _, row in df_character_prompts.iterrows():
    # Preprocess the Prompt
    preprocessed_prompt = preprocess_text(row['prompt'])
    
    # Transform the Prompt into Features
    test_features = vectorizer.transform([preprocessed_prompt])
    
    # Predict
    test_prediction = model.predict(test_features)
    
    # Increment the correct predictions counter if the prediction is 1 (role-play)
    if test_prediction[0] == 1:
        correct_predictions += 1

# Calculate total number of prompts
total_prompts = df_character_prompts.shape[0]

# Calculate accuracy
accuracy_percentage = (correct_predictions / total_prompts) * 100

# Print the total correct predictions and accuracy
print(f"Total Correct Role-Play Predictions: {correct_predictions} out of {total_prompts}")
print(f"Accuracy Percentage: {accuracy_percentage:.2f}%")


Total Correct Role-Play Predictions: 153 out of 153
Accuracy Percentage: 100.00%


In [48]:
test_prompts = [
    "A shadow looms over the city, unnoticed by all but you. What do you do?",
    "You find yourself in the middle of Times Square on New Year's Eve, but everyone has disappeared.",
    "At the stroke of midnight, the book in your hands begins to glow faintly.",
    "Draft an email explaining your strategy for the upcoming marketing campaign, incorporating the theme of 'A Journey Through Time.'",
    "You're at a dinner party, and the person next to you whispers a secret about the host. How do you react?"
]

# Preprocess, transform, and predict for each test prompt
for prompt in test_prompts:
    preprocessed_prompt = preprocess_text(prompt)
    test_features = vectorizer.transform([preprocessed_prompt])
    test_prediction = model.predict(test_features)
    prediction_text = "Character Role-Play" if test_prediction[0] == 1 else "Non-Role-Play"
    print(f"Prompt: {prompt}\nPrediction: {prediction_text}\n")


Prompt: A shadow looms over the city, unnoticed by all but you. What do you do?
Prediction: Non-Role-Play

Prompt: You find yourself in the middle of Times Square on New Year's Eve, but everyone has disappeared.
Prediction: Character Role-Play

Prompt: At the stroke of midnight, the book in your hands begins to glow faintly.
Prediction: Character Role-Play

Prompt: Draft an email explaining your strategy for the upcoming marketing campaign, incorporating the theme of 'A Journey Through Time.'
Prediction: Non-Role-Play

Prompt: You're at a dinner party, and the person next to you whispers a secret about the host. How do you react?
Prediction: Character Role-Play

